File: C#/TheOpenMathLibrary.ActuarialCalculators/BisectionLibrary.cs
namespace ActuarialCalculators
{
    /// <summary>
    /// This BisectionLibrary module in executes a computational technique referred to as the Bisection Method.
    /// </summary>
    public class BisectionLibrary
    {
        /// <summary>
        /// This method determines the root of a formula (ie: where a function reaches zero), i.e., the value of function(x) for which (f(x) = 0).
        /// </summary>
        /// <param name="valueA">The starting point of the interval.</param>
        /// <param name="valueB">The ending point of the interval.</param>
        /// <param name="tolerance">The acceptable error margin for the root.</param>
        /// <param name="maxIterations">The maximum number of iterations to perform.</param>
        /// <param name="mathFunction">The function for which the root is being sought, represented as a delegate Func<double, double>.</param>
        /// <returns></returns>
        /// <exception cref="ArgumentException"></exception>
        public static double Bisection(double valueA, double valueB, double tolerance, int maxIterations, Func<double, double> mathFunction)
        {
            double functionA = mathFunction(valueA);
            double functionB = mathFunction(valueB);
            if (functionA * functionB > 0)
            {
                throw new ArgumentException("function(a) and function(b) must have opposite signs");
            }
            double c = 0;
            for (int i = 0; i < maxIterations; i++)
            {
                c = (valueA + valueB) / 2;
                double functionC = mathFunction(c);
                if (Math.Abs(functionC) < tolerance)
                {
                    break;
                }
                if (functionA * functionC < 0)
                {
                    valueB = c;
                    functionB = functionC;
                }
                else
                {
                    valueA = c;
                    functionA = functionC;
                }
            }
            return c;
        }

        /// <summary>
        /// Newton raphson goalseek method to find the root of a function
        /// </summary>
        /// <param name="x0"></param>
        /// <param name="tolerance"></param>
        /// <param name="maxIterations"></param>
        /// <param name="mathFunction"></param>
        /// <param name="mathFunctionDerivative"></param>
        /// <returns></returns>
        public static double NewtonRaphson(double x0, double tolerance, int maxIterations, Func<double, double> mathFunction, Func<double, double> mathFunctionDerivative)
        {
            double x = x0;
            for (int i = 0; i < maxIterations; i++)
            {
                double functionX = mathFunction(x);
                double derivativeX = mathFunctionDerivative(x);
                double x1 = x - functionX / derivativeX;
                if (Math.Abs(x1 - x) < tolerance)
                {
                    break;
                }
                x = x1;
            }
            return x;
        }

        /// <summary>
        /// Binomial goalseek calculator to find the root of a function
        /// </summary>
        /// <param name="valueA"></param>
        /// <param name="valueB"></param>
        /// <param name="tolerance"></param>
        /// <param name="maxIterations"></param>
        /// <param name="mathFunction"></param>
        /// <returns></returns>
        /// <exception cref="ArgumentException"></exception>
        public static double Binomial(double valueA, double valueB, double tolerance, int maxIterations, Func<double, double> mathFunction)
        {
            double functionA = mathFunction(valueA);
            double functionB = mathFunction(valueB);
            if (functionA * functionB > 0)
            {
                throw new ArgumentException("function(a) and function(b) must have opposite signs");
            }
            double c = 0;
            for (int i = 0; i < maxIterations; i++)
            {
                c = (valueA + valueB) / 2;
                double functionC = mathFunction(c);
                if (Math.Abs(functionC) < tolerance)
                {
                    break;
                }
                if (functionA * functionC < 0)
                {
                    valueB = c;
                    functionB = functionC;
                }
                else
                {
                    valueA = c;
                    functionA = functionC;
                }
            }
            return c;
        }

    }
}

File: Rust/ActuarialAndArithmetic/BisectionLibrary.rs
pub struct BisectionLibrary;

impl BisectionLibrary {
    pub fn bisection<F>(mut value_a: f64, mut value_b: f64, tolerance: f64, max_iterations: usize, math_function: F) -> Result<f64, &'static str>
    where
        F: Fn(f64) -> f64,
    {
        let mut function_a = math_function(value_a);
        let mut function_b = math_function(value_b);
        if function_a * function_b > 0.0 {
            return Err("function(a) and function(b) must have opposite signs");
        }
        let mut c = 0.0;
        for _ in 0..max_iterations {
            c = (value_a + value_b) / 2.0;
            let function_c = math_function(c);
            if function_c.abs() < tolerance {
                break;
            }
            if function_a * function_c < 0.0 {
                value_b = c;
                function_b = function_c;
            } else {
                value_a = c;
                function_a = function_c;
            }
        }
        Ok(c)
    }

    pub fn newton_raphson<F, G>(mut x: f64, tolerance: f64, max_iterations: usize, math_function: F, math_function_derivative: G) -> f64
    where
        F: Fn(f64) -> f64,
        G: Fn(f64) -> f64,
    {
        for _ in 0..max_iterations {
            let function_x = math_function(x);
            let derivative_x = math_function_derivative(x);
            let x1 = x - function_x / derivative_x;
            if (x1 - x).abs() < tolerance {
                break;
            }
            x = x1;
        }
        x
    }

    pub fn binomial<F>(mut value_a: f64, mut value_b: f64, tolerance: f64, max_iterations: usize, math_function: F) -> Result<f64, &'static str>
    where
        F: Fn(f64) -> f64,
    {
        let mut function_a = math_function(value_a);
        let mut function_b = math_function(value_b);
        if function_a * function_b > 0.0 {
            return Err("function(a) and function(b) must have opposite signs");
        }
        let mut c = 0.0;
        for _ in 0..max_iterations {
            c = (value_a + value_b) / 2.0;
            let function_c = math_function(c);
            if function_c.abs() < tolerance {
                break;
            }
            if function_a * function_c < 0.0 {
                value_b = c;
                function_b = function_c;
            } else {
                value_a = c;
                function_a = function_c;
            }
        }
        Ok(c)
    }
}